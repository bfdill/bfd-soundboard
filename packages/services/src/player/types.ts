import { ChildProcess } from 'child_process'

export type DefaultPlayer =
  | 'mplayer'
  | 'afplay'
  | 'mpg123'
  | 'mpg321'
  | 'play'
  | 'omxplayer'
  | 'aplay'
  | 'cmdmp3'
export type What = string | string[]
export type PlayerOptions = {
  player: any
  players?: DefaultPlayer | DefaultPlayer[] | string[]
}
export type PlayerArgs = { [playerName: string]: string[] }
export type PlayOptions = PlayerArgs & { stdio: 'ignore' }
export type Next = Function
export interface IPlayer {
  play: (what: What, options?: PlayOptions | Next, next?: Next) => ChildProcess
}
export type createPlayer = (opts?: PlayerOptions) => IPlayer
