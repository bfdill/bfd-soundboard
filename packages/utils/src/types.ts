export type DeepPartial<T> = {
  [P in keyof T]?: T[P] extends (infer U)[]
    ? DeepPartial<U>[]
    : T[P] extends ReadonlyArray<infer U>
    ? ReadonlyArray<DeepPartial<U>>
    : T[P] extends (string | number | boolean)
    ? T[P]
    : DeepPartial<T[P]>
}
