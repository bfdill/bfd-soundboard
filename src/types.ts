import * as Koa from 'koa'
import * as KoaRouter from 'koa-router'
import { Server } from 'http'

export interface IApplication {
  listen: (port: number) => Server
}

export type ApplicationState = {
  correlationId: string
}
export interface IApplicationContext
  extends Koa.ParameterizedContext<ApplicationState> {}
export type ApplicationMiddleware = Koa.Middleware<
  ApplicationState,
  IApplicationContext
>
export type ApplicationKoa = Koa<ApplicationState, IApplicationContext>
export type ApplicationRouter = KoaRouter<ApplicationState, IApplicationContext>
