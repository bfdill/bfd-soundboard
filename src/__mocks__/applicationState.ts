import { ApplicationState } from '../types'

export const mockApplicationState: ApplicationState = {
  correlationId: 'default_mock_correlation_id'
}
