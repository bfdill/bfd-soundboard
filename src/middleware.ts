import * as compose from 'koa-compose'
import * as winston from 'winston'
import * as uuid from 'uuid/v4'
import { IApplicationContext, ApplicationMiddleware } from './types'

export interface ISystemMiddleware {
  addLoggerToContext: ApplicationMiddleware
  middleware: ApplicationMiddleware
  requestLogger: ApplicationMiddleware
  setInitialState: ApplicationMiddleware
}

export class SystemMiddleware implements ISystemMiddleware {
  constructor(protected readonly logger: winston.Logger) {}

  addLoggerToContext = async (
    context: IApplicationContext,
    next: () => Promise<void>
  ) => {
    context.logger = this.logger
    await next()
  }

  requestLogger = async (
    context: IApplicationContext,
    next: () => Promise<void>
  ) => {
    const start = Date.now()
    let error: any = undefined

    try {
      await next()
    } catch (e) {
      error = e
    }

    const end = Date.now()
    const duration = end - start

    context.logger
      .child({
        middleware: {
          name: 'request logger'
        }
      })
      .log(error === undefined ? 'info' : 'error', 'request', {
        start,
        end,
        duration,
        error
      })
  }

  setInitialState = async (
    context: IApplicationContext,
    next: () => Promise<void>
  ) => {
    context.state = {
      correlationId: uuid()
    }
    context.res.setHeader('x-correlation-id', context.state.correlationId)

    await next()
  }

  middleware = compose<IApplicationContext>([
    this.addLoggerToContext,
    this.setInitialState,
    this.requestLogger
  ])
}
