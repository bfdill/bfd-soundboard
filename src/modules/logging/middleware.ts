import * as uuid from 'uuid/v4'
import { IApplicationContext, ApplicationMiddleware } from '../../types'

export interface ILoggingMiddleware {
  middleware: ApplicationMiddleware
}

export class LoggingMiddleware implements ILoggingMiddleware {
  middleware = (context: IApplicationContext, next: () => Promise<void>) => {
    const correlationId = uuid()
    const start = Date.now()
    let error: any = undefined

    context.state = {
      ...context.state,
      correlationId
    }

    try {
      next()
    } catch (e) {
      error = e
    }

    const end = Date.now()
    const duration = end - start

    context.logger.log(error === undefined ? 'info' : 'error', 'request', {
      start,
      end,
      duration,
      error
    })
  }
}
