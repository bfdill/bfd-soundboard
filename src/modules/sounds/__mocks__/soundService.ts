import { ISoundService, Sound } from '..'

export interface IMockSoundService extends ISoundService {
  setGetBySoundId: (result: Sound | undefined) => void
  setGetSounds: (result: Sound[]) => void
  setIsPathValid: (result: boolean) => void
  getBySoundIdMock: () => jest.Mock
  getSoundsMock: (result: any) => jest.Mock
  IsPathValidMock: (result: boolean) => jest.Mock
}

export const getMockSoundService = (): IMockSoundService => {
  const getBySoundId = jest.fn()
  const getSounds = jest.fn()
  const isPathValid = jest.fn()

  return {
    getBySoundId,
    getSounds,
    isPathValid,
    getBySoundIdMock: getBySoundId,
    getSoundsMock: getSounds,
    IsPathValidMock: isPathValid,
    setGetBySoundId: (result: Sound | undefined) =>
      getBySoundId.mockReturnValue(result),
    setGetSounds: (result: Sound[]) => getSounds.mockReturnValue(result),
    setIsPathValid: (result: boolean) => isPathValid.mockReturnValue(result)
  } as any
}
