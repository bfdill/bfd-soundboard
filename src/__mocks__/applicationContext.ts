import { mockApplicationState } from './applicationState'
import { IApplicationContext } from '../types'
import { getJestLogger } from '../modules/winston-jest/index.test'

export const getMockApplicationContext = (): IApplicationContext =>
  ({
    state: mockApplicationState
  } as any)
