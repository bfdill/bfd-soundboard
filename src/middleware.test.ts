import {
  getJestLogger,
  snapshotExistingLogger,
  IJestLogger
} from './modules/winston-jest/index.test'
import { SystemMiddleware, ISystemMiddleware } from './middleware'
import { IApplicationContext } from './types'
import { getMockApplicationContext } from './__mocks__/applicationContext'
import { inspect } from 'util'

const snapshotLogEntryFailWhaleMethod = (loggedCall: any) => {
  expect(inspect(loggedCall.middleware)).toMatchSnapshot()
  expect(loggedCall.start).toEqual(expect.any(Number))
  expect(loggedCall.end).toEqual(expect.any(Number))
  expect(loggedCall.duration).toEqual(expect.any(Number))
  expect(loggedCall.error).toMatchSnapshot()
  expect(loggedCall.level).toMatchSnapshot()
  expect(loggedCall.message).toMatchSnapshot()
}

describe('middleware', () => {
  let jestLogger: IJestLogger
  let applicationContext: IApplicationContext
  let systemMiddleware: ISystemMiddleware

  beforeEach(() => {
    jestLogger = getJestLogger()
    applicationContext = getMockApplicationContext()
    applicationContext.logger = jestLogger.logger
    systemMiddleware = new SystemMiddleware(jestLogger.logger)
  })

  test('addLoggerToContext', async () => {
    const testLogger = getJestLogger()
    const testContext = getMockApplicationContext()
    const testMiddleware = new SystemMiddleware(testLogger.logger)

    await testMiddleware.addLoggerToContext(testContext, () =>
      Promise.resolve()
    )

    snapshotExistingLogger(testLogger.logger)
  })

  describe('requestLogger', () => {
    test('await next', async () => {
      const next = jest.fn()
      next.mockImplementation(() => Promise.resolve())
      await systemMiddleware.requestLogger(applicationContext, next)
      expect(next).toHaveBeenCalled()

      expect(jestLogger.transport.mock.mock.calls).toHaveLength(1)
      const loggedCall = jestLogger.transport.mock.mock.calls[0][0]
      snapshotLogEntryFailWhaleMethod(loggedCall)
      jestLogger.transport.mock.mockReset()
    })

    test('doh error - muh bad', async () => {
      const next = jest.fn()
      next.mockRejectedValue('muh Bad')
      await systemMiddleware
        .requestLogger(applicationContext, next)
        .catch((reason: any) => {
          expect(reason).toMatchSnapshot()
        })

      expect(jestLogger.transport.mock.mock.calls).toHaveLength(1)
      const loggedCall = jestLogger.transport.mock.mock.calls[0][0]
      snapshotLogEntryFailWhaleMethod(loggedCall)
      jestLogger.transport.mock.mockReset()
    })
  })

  test('setInitialState', async () => {
    const setHeader = jest.fn()
    const context: IApplicationContext = {
      res: { setHeader }
    } as any

    await systemMiddleware.setInitialState(context, () => Promise.resolve())

    expect(context.state).toBeDefined()
    expect(context.state.correlationId).toBeDefined()
    expect(setHeader).toBeCalledWith(
      'x-correlation-id',
      context.state.correlationId
    )
    jestLogger.callsMatchSnapshot()
  })
})
