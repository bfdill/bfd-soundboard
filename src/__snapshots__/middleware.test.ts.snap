// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`middleware addLoggerToContext 1`] = `
Array [
  Object {
    "level": "info",
    "message": "✅test works✅",
    Symbol(level): "info",
    Symbol(message): "{\\"message\\":\\"✅test works✅\\",\\"level\\":\\"info\\"}",
  },
]
`;

exports[`middleware requestLogger await next 1`] = `"{ name: 'request logger' }"`;

exports[`middleware requestLogger await next 2`] = `undefined`;

exports[`middleware requestLogger await next 3`] = `"info"`;

exports[`middleware requestLogger await next 4`] = `"request"`;

exports[`middleware requestLogger doh error - muh bad 1`] = `"{ name: 'request logger' }"`;

exports[`middleware requestLogger doh error - muh bad 2`] = `"muh Bad"`;

exports[`middleware requestLogger doh error - muh bad 3`] = `"error"`;

exports[`middleware requestLogger doh error - muh bad 4`] = `"request"`;
